src/witness/oracle.rs

```diff
@@ -552,8 +552,35 @@ pub fn create_artifacts_from_tracer<
         // wherever we have this marker we should look at the tail of the item right before it
         let pos = log_position_mapping[&rollback_tail_marker];
         let tail = if pos == -1 {
-            // empty
-            global_end_of_storage_log
+            if frame_index > 1 {
+                // empty frame
+                let pos_forward_head =
+                    log_position_mapping[&ExtendedLogQuery::FrameForwardHeadMarker(frame_index)];
+                let pos_forward_tail =
+                    log_position_mapping[&ExtendedLogQuery::FrameForwardTailMarker(frame_index)];
+                let pos_rollback_head =
+                    log_position_mapping[&ExtendedLogQuery::FrameRollbackHeadMarker(frame_index)];
+
+                if pos_forward_head == -1 && pos_forward_tail == -1 && pos_rollback_head == -1 {
+                    // absolutely empty frame, most likely leaf one,
+                    // but we can not just use global end,
```

Comment:
why we cannot use global end?

Reply:
Because we _can_ have some state changes after this rollback. But our rollback tail should be equal to the last state change tail before the rollback.

Global end was used under assumption that situation without rollback marker can happen only in trivial cases (like entry bytecode without any state changes)